lucille-plugins/lucille-pinecone/src/main/java/com/kmwllc/lucille/pinecone/indexer/PineconeIndexer.java

```diff
@@ -33,24 +55,33 @@ public class PineconeIndexer extends Indexer {
   private final Set<String> metadataFields;
   private final String mode;
   private final String defaultEmbeddingField;
+  private final String addDeletionPrefix;
+  private final boolean deleteByPrefix;
 
   public PineconeIndexer(Config config, IndexerMessenger messenger, String metricsPrefix) {
     super(config, messenger, metricsPrefix);
     this.indexName = config.getString("pinecone.index");
     this.namespaces = config.hasPath("pinecone.namespaces") ? config.getConfig("pinecone.namespaces").root().unwrapped() : null;
-    this.metadataFields = new HashSet<>(config.getStringList("pinecone.metadataFields"));
+    this.metadataFields = config.hasPath("pinecone.metadataFields")
+        ? new HashSet<>(config.getStringList("pinecone.metadataFields")) : new HashSet<>();
     this.mode = config.hasPath("pinecone.mode") ? config.getString("pinecone.mode") : "upsert";
     this.defaultEmbeddingField = ConfigUtils.getOrDefault(config, "pinecone.defaultEmbeddingField", null);
     this.client = new Pinecone.Builder(config.getString("pinecone.apiKey")).build();
     this.index = this.client.getIndexConnection(this.indexName);
-    if (namespaces == null && defaultEmbeddingField == null) {
+    this.addDeletionPrefix = config.hasPath("pinecone.addDeletionPrefix") ? config.getString("pinecone.addDeletionPrefix") : "";
+    this.deleteByPrefix = config.hasPath("pinecone.deleteByPrefix") ? config.getBoolean("pinecone.deleteByPrefix") : false;
+    if (namespaces == null && defaultEmbeddingField == null && !mode.equalsIgnoreCase("delete")) {
```

Comment:
in "delete" mode, it is not necessary to have namespaces or defaultEmbeddingField. This is because we allow upserting and updating when namespaces is null and defaultEmbeddingField is set. In this case, vectors will be uploaded to "default" namespace. "delete" mode will then delete from "default" if namespaces is null

Reply:
no more "delete" mode. Instead moved validation part to when there is a need to upload documents